bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java

```diff
@@ -141,22 +142,53 @@ public String toString() {
         return classModel;
     }
 
+
+    private LevelCountingBsonWriter wrapBsonWriter(final BsonWriter writer) {
+        if (writer instanceof LevelCountingBsonWriter) {
+            return (LevelCountingBsonWriter) writer;
+        }
+        return new LevelCountingBsonWriter(writer);
+    }
+
+    private <S> void encodeIdProperty(final LevelCountingBsonWriter writer, final T instance, final EncoderContext encoderContext,
+                                      final IdPropertyModelHolder<S> propertyModelHolder) {
+        if (propertyModelHolder.getPropertyModel() == null || propertyModelHolder.getIdGenerator() == null) {
+            encodeProperty(writer, instance, encoderContext, propertyModelHolder.getPropertyModel());
+        } else {
+            S id = propertyModelHolder.getPropertyModel().getPropertyAccessor().get(instance);
+            if (id == null && writer.getCurrentLevel() == 0 && encoderContext.isEncodingCollectibleDocument()) {
```

Comment:
Now that it's using EncoderContext.isEncodingCollectibleDocument, it's no longer necessary to use a LevelCountingBsonWriter.  It just has to use EncoderContext.encodeWithChildContext on line 18 of encodeValue, below (similar to how DocumentCodec, etc. work)

Reply:
Much better - Done